### DISCUSSION TITLE
Directories option bdist_msi

### DISCUSSION THREAD

**Discussion by DomeliPierre** [marcelotduarte/cx_Freeze#1043] [Q&A]
Hi, What is the exact syntax for the directories option, what is the meaning of each item of the 3-tuple ? 
I try : 
'directories' : [('MyAppliDir', '[ProgramData]', "temp")],  I get MSI error 2705 at the start of execution of the installer
'directories' : [('MyAppliDir', '', "temp")],  I get MSI error 2704
'directories' : [('MyAppliDir', 'c:\\', "c:\\temp")],  I get MSI error 2704
Thanks for your help

**Comment by marcelotduarte** [comment]
bdist_msi uses [msilib](https://docs.python.org/3/library/msilib.html?highlight=directory#module-msilib)
In https://docs.python.org/3/library/msilib.html#directory-objects has reference to https://msdn.microsoft.com/en-us/library/windows/desktop/aa368295.aspx

(It is not an answer, but it is an attempt to help you find the answer.)

**Reply by DomeliPierre** [reply]
According to the Microsoft's doc found in 
https://docs.microsoft.com/en-us/windows/win32/msi/createfolder-table
and https://docs.microsoft.com/en-us/windows/win32/msi/component-table
we must tell msi to not delete at the end of the installation the user's directory we just created using the 'directories' option. It is done by adding a line in the CreateFolder table:
```
v2 = db.OpenView("SELECT * FROM `%s`" % 'CreateFolder')
count2 = v2.GetColumnInfo(MSICOLINFO_NAMES).GetFieldCount()
r2=CreateRecord(count2)
r2.SetString(1,value[0]);  # value[0] keeps the user's directory name
r2.SetString(2,value[0])  # External key into the first column of the Directory table.
v2.Modify(MSIMODIFY_INSERT, r2) 
r2.ClearData()

# now we have to create a link with the component table
component = value[0]
uuid = gen_uuid()
logical = value[0]
flags = 256
keyid = None
add_data(db, "Component", [(component, uuid, logical, flags, None, keyid)])

The setup used to create user's (empty) directories during the installation is : 
setup(
    name="test",
    version="0.1",
    description="script for testing bdist-msi directories option",
    executables=[Executable("Hello.py")],
    options={
        "build_exe": { 'silent':True },
        "bdist_msi": { 'directories': [
                                        ('tempfolder1', 'TARGETDIR', "TEMP~1|tempfolder1"), 
                                        ('tempfolder2', 'tempfolder1', "TEMP~1|tempfolder2"), 
                                        ('CommonAppDataFolder', None, 'SourceDir'),
                                        ('tempfolder3', 'CommonAppDataFolder', "TEMP~1|tempfolder3"),
                                      ] 
                     }
    }
)
```
There is no Windows Installer error message during the installation, but something is missing because the expected result isn't there.
Does someone have an idea ?

**Reply by marcelotduarte** [reply]
I found some documentation in an issue and updated the example of bdist_msi:
https://cx-freeze.readthedocs.io/en/latest/distutils.html#bdist-msi

**Reply by marcelotduarte** [reply]
cx_Freeze 6.7 has just been released.
